replica/database.cc

```diff
@@ -2833,29 +2800,70 @@ future<> database::drain() {
     b.cancel();
 }
 
+void database::tables_metadata::add_table_helper(database& db, keyspace& ks, table& cf, schema_ptr s) {
+    // A table needs to be added atomically.
+    auto id = s->id();
+    ks.add_or_update_column_family(s);
+    auto remove_cf1 = defer([&] () noexcept { ks.metadata()->remove_column_family(s); });
+    s->registry_entry()->set_table(cf.weak_from_this());
+
+    _column_families.emplace(id, s->table().shared_from_this());
+    auto remove_cf2 = defer([&] () noexcept {
+        _column_families.erase(s->id());
+    });
+    _ks_cf_to_uuid.emplace(std::make_pair(s->ks_name(), s->cf_name()), id);
+    auto remove_cf3 = defer([&] () noexcept {
+        _ks_cf_to_uuid.erase(std::make_pair(s->ks_name(), s->cf_name()));
+    });
+
+    if (s->is_view()) {
+        db.find_column_family(s->view_info()->base_id()).add_or_update_view(view_ptr(s));
+    }
+    auto remove_view = defer([&] () noexcept {
+        if (s->is_view()) {
+            try {
+                db.find_column_family(s->view_info()->base_id()).remove_view(view_ptr(s));
+            } catch (no_such_column_family&) {
+                // Drop view mutations received after base table drop.
+            }
+        }
+    });
```

Comment:
The last undo is unneeded because there's nothing left that can fail. I don't mind keeping it here in case another item is added.